src/proxy.c

```diff
@@ -30,108 +39,234 @@ SEXP vec_proxy(SEXP x) {
   return out;
 }
 
+SEXP vec_proxy_method(SEXP x) {
+  return s3_find_method("vec_proxy", x, vctrs_method_table);
+}
+
+// This should be faster than normal dispatch but also means that
+// proxy methods can't call `NextMethod()`. This could be changed if
+// it turns out a problem.
+SEXP vec_proxy_invoke(SEXP x, SEXP method) {
+  if (method == R_NilValue) {
+    return x;
+  } else {
+    return vctrs_dispatch1(syms_vec_proxy, method, syms_x, x);
+  }
+}
+
+
+static inline SEXP vec_proxy_equal_method(SEXP x);
+static inline SEXP vec_proxy_equal_invoke(SEXP x, SEXP method);
+
 // [[ register(); include("vctrs.h") ]]
 SEXP vec_proxy_equal(SEXP x) {
-  SEXP proxy = PROTECT(vec_proxy_recursive(x, vctrs_proxy_equal));
+  SEXP method = PROTECT(vec_proxy_equal_method(x));
+  SEXP out = vec_proxy_equal_invoke(x, method);
+  UNPROTECT(1);
+  return out;
+}
 
-  if (is_data_frame(proxy)) {
-    // Flatten df-cols so we don't have to recurse to work with data
-    // frame proxies
-    proxy = PROTECT(df_flatten(proxy));
+static inline SEXP vec_proxy_equal_method(SEXP x) {
+  SEXP cls = PROTECT(s3_get_class(x));
+  SEXP method = s3_class_find_method("vec_proxy_equal", cls, vctrs_method_table);
 
-    // Unwrap data frames of size 1 since the data frame wrapper
-    // doesn't impact rowwise equality or identity
-    proxy = vec_proxy_unwrap(proxy);
+  if (method != R_NilValue) {
+    UNPROTECT(1);
+    return method;
+  }
 
+  // FIXME: Stopgap check for bare arrays, which equality functions don't
+  // handle well
+  if (vec_dim_n(x) > 1) {
     UNPROTECT(1);
+    return fns_vec_proxy_equal_array;
   }
 
   UNPROTECT(1);
-  return proxy;
+  return R_NilValue;
 }
-static SEXP vec_proxy_unwrap(SEXP x) {
-  if (TYPEOF(x) == VECSXP && XLENGTH(x) == 1 && is_data_frame(x)) {
-    x = vec_proxy_unwrap(VECTOR_ELT(x, 0));
+
+static inline SEXP vec_proxy_equal_invoke(SEXP x, SEXP method) {
+  if (method != R_NilValue) {
+    return vctrs_dispatch1(syms_vec_proxy_equal, method, syms_x, x);
+  }
+
+  // Fallback to `vec_proxy()` on S3 objects with no equality proxy
+  if (vec_typeof(x) == vctrs_type_s3) {
+    return vec_proxy(x);
+  } else {
+    return x;
   }
-  return x;
 }
 
-// [[ register() ]]
-SEXP vctrs_unset_s4(SEXP x) {
-  x = r_clone_referenced(x);
-  r_unmark_s4(x);
-  return x;
+
+static inline SEXP vec_proxy_compare_method(SEXP x);
+static inline SEXP vec_proxy_compare_invoke(SEXP x, SEXP method);
+
+// [[ register(); include("vctrs.h") ]]
+SEXP vec_proxy_compare(SEXP x) {
+  SEXP method = PROTECT(vec_proxy_compare_method(x));
+  SEXP out = vec_proxy_compare_invoke(x, method);
+  UNPROTECT(1);
+  return out;
+}
+
+static inline SEXP vec_proxy_compare_method(SEXP x) {
```

Comment:
If possible it would be nice to remove the code repetition, this way we see there is no difference.